src/tritongrader/test_case.py

```diff
@@ -223,3 +255,55 @@ def _generate_summary(self, verbose=False):
 
     def generate_test_summary(self, verbose=False):
         return self._generate_summary(verbose)
+
+
+class CustomTestCaseResult(TestCaseResultBase):
+    def __init__(self):
+        super().__init__()
+        self.output: str = ""
+
+
+class CustomTestCase(TestCaseBase):
+    def __init__(
+        self,
+        func: Callable[[CustomTestCaseResult], None],
+        name: str = "Test Case",
+        point_value: float = 1,
+        timeout: float = TestCaseBase.DEFAULT_TIMEOUT_SECS,
+        visibility: Visibility = Visibility.VISIBLE,
+    ):
+        super().__init__(name, point_value, timeout, visibility)
+        self.test_func: Callable[[CustomTestCaseResult], bool] = func
+
+    def execute(self):
+        self.result = CustomTestCaseResult()
+        self.result.has_run = True
+
+        try:
+            t = threading.Thread(
+                target=self.test_func,
+                args=[self.result],
+            )
+            t.start()
+            t.join(timeout=self.timeout)
+            if t.is_alive():
+                raise TimeoutError
+        except TimeoutError:
+            logger.info(f"{self.name} timed out (limit={self.timeout}s)!")
```

Comment:
Does this kill the thread? If not, keeping the CPU hogged isn't great. In that case the thread should take care of timeout internally.

Reply:
No this does not kill the thread. Afaik there's no elegant way in Python to programmatically kill a thread.

So yes the thread would need to take care of timeout internally. That's probably a safer way to do it too, since the parent thread killing a child thread could have unintended consequences especially with shared data structures involved (i.e., the results object).